### PR TITLE
[DTensor] statically_known_true for slice strategy

### PR DESCRIPTION
**Pull request by pianpwk**
Avoids data-dependent errors for out-of-bounds & redundant slice checks.

The sharding logic that immediately depends on this only checks for redundant slices, and is saying: "it's safe to reuse the input placements if a) the slicing dimension isn't sharded, or b) the slice is redundant, so just pretend this op didn't happen".

This has a slight effect on output placements, when a slice is performed on a shared dim, and dynamic shapes are involved (size/start/end/step). Now if the slice isn't obviously redundant, we won't immediately consider the input placements valid (even if they could be for very particular runtime shapes), and select strategies valid for the general case - in this case I guess unsharding the slicing dim.

For backed symbols, we could choose to recompile when the redundant case is hit, by switching to `guard_or_false`, but it's not obvious how desirable this is.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #166990

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

File: test/distributed/tensor/test_dtensor_export.py [hunk 1]
```diff
@@ -548,6 +548,16 @@ def forward(self, x, y):
...
 
+        class Bar(torch.nn.Module):
+            def forward(self, x):
+                val = torch.clamp(x.max(), min=1).item()
+                torch._check(val >= 1)
+                return x[:val]
+
+        x = torch.randint(1000, (4, 64, 16))
+        x_dt = distribute_tensor(x, device_mesh, placements=[Replicate()])
+        _dynamo_graph_capture_for_export(Bar())(x_dt)
```
Comment (tugsbayasgalan) [inline test/distributed/tensor/test_dtensor_export.py]: maybe assert the graph?

**Comment by ezyang** [comment]
This is kind of spooky. It might be right, but I'm worried. Is this showing up in one of the models we care about?

**Comment by pianpwk** [comment]
> This is kind of spooky. It might be right, but I'm worried. Is this showing up in one of the models we care about?

hmm @tugsbayasgalan where was this coming from?

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 2]
```diff
@@ -381,7 +382,7 @@ def gen_slice_strategy(op_schema: OpSchema) -> StrategyType:
         raise AssertionError(f"Expected int, got {type(dim)}")
     if start is None:
         start = 0
-    if end is None or end > input_shape[dim]:
+    if end is None or statically_known_true(end > input_shape[dim]):
```
Comment (ezyang) [inline torch/distributed/tensor/_ops/_tensor_ops.py:385]: You need to check if this results in incorrect behavior when the end slice is actually out of bounds for the tensor at runtime; I can't easily tell. What's difficult to understand is why the sharding rule needs such fine grained information. You would not need to change this to statically known true if you had an extra assert that the value is in bounds for the tensor end, which I believe is true in the real example.

File: torch/distributed/tensor/_ops/_tensor_ops.py [hunk 3]
```diff
@@ -395,13 +396,20 @@ def gen_slice_strategy(op_schema: OpSchema) -> StrategyType:
...
 
     slice_strategy = OpStrategy([])
 
     for arg_strategy in input_strategy.strategies:
         arg_spec = arg_strategy.output_spec
-        if not is_tensor_dim_sharded(arg_spec, dim=slice_dim) or redundant_slice:
+        if (
+            not is_tensor_dim_sharded(arg_spec, dim=slice_dim)
+            or statically_redundant_slice
+        ):
```
Comment (ezyang) [inline torch/distributed/tensor/_ops/_tensor_ops.py:412]: Ah, this is indeed safe to skip. But then you can simplify most of the logic above: essentially you are trying to see if the slice is redundant, and it is always ok to give up and then have less shard rules

**Review by ezyang — COMMENTED** [review]
I am going to let Laith make the call but upon closer inspection this looks ok.

**Review by laithsakka — APPROVED** [review]
this mach the c++,
For future items it might be good to set an invariant that that backed matches eager when it comes to sharding decisions.
